From 4793fd705e5f83d6abae2d644007c541e85ce710 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Wed, 28 Jun 2023 14:45:27 +0200
Subject: [PATCH 0561/1695] drivers: stm32_rifsc: add RIMU logic
 reconfiguration API

Adds stm32_rifsc_reconfigure_rimu() API that allows a RIMU
reconfiguration at run time.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 53a1ef777fe60a9ad8cf3ed48c77d0c2a775b612 drivers: stm32_rifsc: add RIMU logic reconfiguration API]
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: I5812308707cf5978fa3721535d3927c428c7c52f
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/354893
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 core/drivers/firewall/stm32_rifsc.c | 40 +++++++++++++++++++++++++++++
 core/include/drivers/stm32_rif.h    | 19 ++++++++++++++
 2 files changed, 59 insertions(+)

diff --git a/core/drivers/firewall/stm32_rifsc.c b/core/drivers/firewall/stm32_rifsc.c
index 3d8037b85..46e395122 100644
--- a/core/drivers/firewall/stm32_rifsc.c
+++ b/core/drivers/firewall/stm32_rifsc.c
@@ -515,6 +515,46 @@ static TEE_Result stm32_rimu_setup(struct rifsc_platdata *pdata)
 	return 0;
 }
 
+TEE_Result stm32_rifsc_reconfigure_rimu(unsigned int id,
+					unsigned int master_cid,
+					bool cid_sel, bool sec, bool priv)
+{
+	struct rimu_cfg *rimu = NULL;
+	uint32_t rimu_conf = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	if (id > rifsc_pdata.drv_data->nb_rimu ||
+	    master_cid > MAX_CID_SUPPORTED)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	rimu = &rifsc_pdata.rimu[id];
+
+	if (io_read32(rifsc_pdata.base + _RIFSC_RIMC_CR) &
+	    _RIFSC_RIMC_CR_GLOCK) {
+		DMSG("RIMU configuration is locked");
+		return TEE_ERROR_ACCESS_DENIED;
+	}
+
+	if (cid_sel)
+		rimu_conf |= RIFSC_RIMC_MODE_MASK;
+	if (sec)
+		rimu_conf |= RIFSC_RIMC_MSEC_MASK;
+	if (priv)
+		rimu_conf |= RIFSC_RIMC_MPRIV_MASK;
+
+	rimu_conf |= master_cid << RIFSC_RIMC_MCID_SHIFT;
+
+	rimu->attr = rimu_conf;
+
+	res = stm32_rimu_cfg(&rifsc_pdata, rimu);
+	if (res) {
+		EMSG("RIMU %u reconfiguration error", id);
+		return res;
+	}
+
+	return TEE_SUCCESS;
+}
+
 static TEE_Result stm32_rifsc_sem_pm_suspend(void)
 {
 	int i = 0;
diff --git a/core/include/drivers/stm32_rif.h b/core/include/drivers/stm32_rif.h
index 99ad6881e..88348d2dc 100644
--- a/core/include/drivers/stm32_rif.h
+++ b/core/include/drivers/stm32_rif.h
@@ -97,6 +97,25 @@ static inline TEE_Result stm32_rifsc_check_tdcid(bool *tdcid_state)
 }
 #endif
 
+/*
+ * Reconfigures a specific RIMU
+ *
+ * @id:		ID of the RIMU to reconfigure.
+ * @master_cid: CID that will be linked to the RIMU.
+ * @cid_sel:	Drives the inheritance of the CID of the RIMU.
+ * @sec:	Master is secure/non-secure
+ * @priv:	Master is privileged/non-privileged
+ *
+ * Returns TEE_SUCCESS in case of success.
+ * Returns TEE_ERROR_BAD_PARAMETERS if either the RIMU ID or the master cid is
+ * incorrect.
+ * Returns TEE_ERROR_ACCESS_DENIED if the RIMU can't be reconfigured.
+ * Returns any TEE_Result compliant code in case of error.
+ */
+TEE_Result stm32_rifsc_reconfigure_rimu(unsigned int id,
+					unsigned int master_cid,
+					bool cid_sel, bool sec, bool priv);
+
 /*
  * Check every possible configuration where accessing RIF
  * is possible :
-- 
2.47.2

